### PR TITLE
Support nested URL macros

### PR DESCRIPTION
**Pull request by calebcordry**
This is the start to a process that should culminate with the new added url variable filtering functionality Related-to #2198 

This introduces the new parser that will live in core. This step is taken so that these macros can all be resolved concurrently, which will eventually allow more complex operations to be resolved in urls for analytics.

### CODE REVIEW

File: src/service/url-expander/parser.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+   * excludes protected keywords.
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {string} url The url to be expanded.
+   * @return {!Object<string, string|number>}
+   */
+  eliminateOverlaps_(matches, url, opt_whiteList) {
+    if (!matches.length) { return null; }
+    // longest keywords have priority over shorter
+    // how should we prioritize if same length??
```
Comment (calebcordry) [inline src/service/url-expander/parser.js]: Right now this is kind of arbitrary, should we have a way to decide if keywords are same length?

File: src/service/url-replacements-impl.js [hunk 2]
```diff
@@ -934,6 +937,13 @@ export class UrlReplacements {
    * @private
    */
   expand_(url, opt_bindings, opt_collectVars, opt_sync, opt_whiteList) {
```
Comment (calebcordry) [inline src/service/url-replacements-impl.js]: There are some legacy tests of this that will fail with the new parser. Not sure if its better to refactor tests or code for some of them. @lannka we should discuss next week

File: src/service/url-expander/parser.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+          }
+          builder = '';
+          urlIndex++;
+
+        }
+
+        else if (url[urlIndex] === ')') {
+          urlIndex++;
+          const binding = stack.pop();
+          const args = [...results, builder.trim()];
```
Comment (calebcordry) [inline src/service/url-expander/parser.js]: Right now we are trimming all args to support `arg1, arg2` but we may choose to drop this if it increases complexity

File: test/functional/url-expander/test-parser.js [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+          .to.eventually.equal('aaaBBBccc')
+    );
+
+    it('should handle real urls', () => {
+      const url = 'http://www.amp.google.com/?client=CLIENT_ID(__ga)&canon=CANONICAL_URL&random=RANDOM';
+      const expected = 'http://www.amp.google.com/?client=amp-GA12345&canon=www.google.com&random=123456';
+      return expect(parser.expand(url, mockBindings)).to.eventually.equal(expected);
+    });
+
+    it.skip('parses one unknown function, one argument', () =>
```
Comment (calebcordry) [inline test/functional/url-expander/test-parser.js]: these are skipped for now until best way to treat unknown keywords is discussed further

File: src/service/url-expander/parser.js [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+   *  matching keywords
+   * @param {string} url The url to be expanded.
+   * @return {!Object<string, string|number>}
+   */
+  eliminateOverlaps_(matches, url, opt_whiteList) {
+    if (!matches.length) { return null; }
+    // longest keywords have priority over shorter
+    // how should we prioritize if same length??
+    matches.sort((a, b) => b.length - a.length);
+    const storage = new Array(url.length).fill(false);
```
Comment (lannka) [inline src/service/url-expander/parser.js]: storage is too generic term. let's call it `keywordMarker` or so.

File: src/service/url-expander/parser.js [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+  /**
+   * Takes array of all matching variable keyword locations and returns only those that we
+   * want to evaluate. In the case of overlapping keywords it choose the longer. It also
+   * excludes protected keywords.
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {string} url The url to be expanded.
+   * @return {!Object<string, string|number>}
+   */
+  eliminateOverlaps_(matches, url, opt_whiteList) {
```
Comment (lannka) [inline src/service/url-expander/parser.js]: order the methods in such a way:

1. public methods first
2. constructor first
3. sort private methods in the order of their invocations
4. if A invokes B, put A before B.

File: src/service/url-expander/expander.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    }, []).sort((a, b) => a.start - b.start);
+  }
+
+  /**
+   * Given a keyword marking data structure and a match object, it will keep track of
+   * indexes that are already being tracked.
+   * @param {!Object<string, *>} match contains match data
+   * @return {boolean} indicates whether a collision was found
+   */
+  fillStorage_(match, keywordMarker) {
```
Comment (lannka) [inline src/service/url-expander/expander.js]: fillStorage_ => markKeywords_

File: src/service/url-expander/expander.js [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    matches.sort((a, b) => {
+      if (a.length === b.length) {
+        return a.name > b.name;
+      }
+      return b.length - a.length;
+    });
+
+    const keywordMarker = new Array(url.length).fill(false);
+
+    return matches.reduce((results, match) => {
```
Comment (lannka) [inline src/service/url-expander/expander.js]: In this case, isn't [Array.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) easier to read and write than `Array.reduce`?

File: src/service/url-expander/expander.js [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {string} url The url to be expanded.
+   * @return {!Object<string, string|number>}
+   */
+  eliminateOverlaps_(matches, url, opt_whiteList) {
+    if (!matches.length) { return null; }
+    // longest keywords have priority over shorter
+    // if same length alphabetical precidence
+    matches.sort((a, b) => {
```
Comment (lannka) [inline src/service/url-expander/expander.js]: might worth to extract the sorting function to constants.

File: src/service/url-expander/expander.js [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+          }
+
+          builder = '';
+        }
+
+        else if (numOfPendingCalls && url[urlIndex] === ',') {
+          if (builder.length) {
+            results.push(builder.trim());
+          }
+          // support legacy two comma format
```
Comment (lannka) [inline src/service/url-expander/expander.js]: shouldn't be called legacy, i guess that's still the way to do it now, right?

File: src/service/url-expander/expander.js [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+  /**
+   * Given a keyword marking data structure and a match object, it will keep track of
+   * indexes that are already being tracked.
+   * @param {!Object<string, *>} match contains match data
+   * @return {boolean} indicates whether a collision was found
+   */
+  fillStorage_(match, keywordMarker) {
+    for (let i = match.start; i <= match.stop; i++) {
+      if (keywordMarker[i]) {
+        // toggle off what we thought was to be filled
```
Comment (lannka) [inline src/service/url-expander/expander.js]: ha, now I think the other algorithm is cheaper:

```js
for (let i = 0; i < usedRanges.length; i++) {
  if (hasOverlap(match, usedRanges[i])) {return false;}
});
return true;

```

File: src/service/url-expander/expander.js (hunk 10 above)
Comment (calebcordry) [reply src/service/url-expander/expander.js]: yeah we have to keep this to not break things. Should I just call it existing?

File: src/service/url-expander/expander.js (hunk 11 above)
Comment (calebcordry) [reply src/service/url-expander/expander.js]: the algorithm for merging the overlaps? or the other expanding algorithm?

Comment (calebcordry) [reply src/service/url-expander/expander.js]: never mind, just figured out what you are saying

File: src/service/url-expander/expander.js (hunk 8 above)
Comment (calebcordry) [reply src/service/url-expander/expander.js]: yes filter is better 👍

File: src/service/url-replacements-impl.js [hunk 12]
```diff
@@ -52,7 +54,7 @@ const ORIGINAL_VALUE_PROPERTY = 'amp-original-value';
  * @param {*} val
  * @return {string}
  */
-function encodeValue(val) {
+export function encodeValue(val) {
```
Comment (lannka) [inline src/service/url-replacements-impl.js]: is this used outside?

File: src/service/url-replacements-impl.js [hunk 13]
```diff
@@ -639,6 +639,9 @@ export class UrlReplacements {
 
     /** @type {VariableSource} */
     this.variableSource_ = variableSource;
+
+    /** @type {Expander} */
```
Comment (lannka) [inline src/service/url-replacements-impl.js]: `{!Expander}`, since it will not be `null`

File: src/service/variable-source.js [hunk 14]
```diff
@@ -165,6 +168,7 @@ export class VariableSource {
         this.replacements_[varName] || {sync: undefined, async: undefined};
     this.replacements_[varName].sync = syncResolver;
     this.replacementExpr_ = undefined;
+    this.replacementExprWithoutArgs_ = undefined;
```
Comment (lannka) [inline src/service/variable-source.js]: maybe name it `replacementExprV2_`, so we know the other one is going to be remove after v2 launch.

in general, we can have consistent naming (suffix of v2) so the later clean up can be easier.

File: src/service/variable-source.js [hunk 15]
```diff
@@ -217,20 +223,30 @@ export class VariableSource {
...
   }
 
   /**
    * @param {!Array<string>} keys
+   * @param {boolean=} opt_ignoreArgs flag to ignore capture of args
    * @return {!RegExp}
    * @private
    */
-  buildExpr_(keys) {
+  buildExpr_(keys, opt_ignoreArgs) {
```
Comment (lannka) [inline src/service/variable-source.js]: `opt_ignoreArgs` -> `isV2`

File: test/functional/url-expander/test-expander.js [hunk 16]
```diff
@@ -0,0 +1,870 @@
...
+    sandbox.restore();
+  });
+
+  describe('#eliminateOverlaps', () => {
+    const url = 'http://www.google.com/?client=CLIENT_ID(__ga)&canon=CANONICAL_URL&random=RANDOM';
+
+    it('should handle empty', () => {
+      const array = [];
+      const expected = null;
+      expect(expander.eliminateOverlaps_(array, url)).to.deep.equal(expected);
```
Comment (lannka) [inline test/functional/url-expander/test-expander.js]: we prefer to test public method only, so if later on we changed the internal implementation but not the public API, we're confident if it passes the current tests.

in this case, you need to come up with some test case that have overlaps in the URL, and test the end result of `expand()`

File: test/functional/url-expander/test-expander.js [hunk 17]
```diff
@@ -0,0 +1,870 @@
...
+    });
+
+    it('should replace STORY_PAGE_INDEX and STORY_PAGE_ID' +
+        ' with empty string if amp-story is not configured', () => {
+      return expect(
+          expander.expand('?index=STORY_PAGE_INDEX&id=STORY_PAGE_ID'))
+          .to.eventually.equal('?index=&id=');
+    });
+
+    it('should replace TIMESTAMP', () => {
```
Comment (lannka) [inline test/functional/url-expander/test-expander.js]: Maybe I didn't make it clear. Those variables are not implemented in this expander, hence the tests should not go here.

They should be in `url-replacement-impl-test.js`. If we start to extract GlobalVariableSource to a separate file, then we can consider have them in `test-global-variable-source.js` at that time.

File: src/service/url-expander/expander.js (hunk 10 above)
Comment (lannka) [reply src/service/url-expander/expander.js]: just remove the wording `legacy`, cause it's still by design

File: src/service/url-expander/parser.js [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {rethrowAsync} from '../../log';
+import {encodeValue} from '../url-replacements-impl';
+
+/** Rudamentary parser to handle nested Url replacement. */
+export class Parser {
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply src/service/url-expander/parser.js]: changed to `Expander`

File: src/service/url-expander/parser.js (hunk 6 above)
Comment (calebcordry) [reply src/service/url-expander/parser.js]: reordered

File: src/service/url-expander/parser.js (hunk 5 above)
Comment (calebcordry) [reply src/service/url-expander/parser.js]: changed to keywordMarker

File: tools/experiments/experiments.js [hunk 19]
```diff
@@ -287,6 +287,12 @@ const EXPERIMENTS = [
     spec: 'https://github.com/ampproject/amphtml/issues/6469',
     cleanupIssue: 'https://github.com/ampproject/amphtml/issues/12267',
   },
+  {
+    id: 'core-parser',
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply tools/experiments/experiments.js]: changed experiment name to url-replacement-v2

File: src/service/url-expander/expander.js (hunk 9 above)
Comment (calebcordry) [reply src/service/url-expander/expander.js]: irrelevant since we confirmed regex behavior

File: src/service/url-expander/expander.js (hunk 8 above)
Comment (calebcordry) [reply src/service/url-expander/expander.js]: changed to filter

File: src/service/url-expander/expander.js (hunk 7 above)
Comment (calebcordry) [reply src/service/url-expander/expander.js]: changed to markKeywords_

File: src/service/url-expander/expander.js (hunk 11 above)
Comment (calebcordry) [reply src/service/url-expander/expander.js]: part of deleted sorting code

File: src/service/url-replacements-impl.js (hunk 12 above)
Comment (calebcordry) [reply src/service/url-replacements-impl.js]: I was using this at first, but deleted in a refactor, and forgot to clean this up. Export is removed.

File: src/service/url-replacements-impl.js (hunk 13 above)
Comment (calebcordry) [reply src/service/url-replacements-impl.js]: changed type

File: test/functional/url-expander/test-expander.js (hunk 16 above)
Comment (calebcordry) [reply test/functional/url-expander/test-expander.js]: we discussed this one in person, but tests were changed. Also pretty much made irrelevant with regex behavior

File: test/functional/url-expander/test-expander.js (hunk 17 above)
Comment (calebcordry) [reply test/functional/url-expander/test-expander.js]: removed most of the tests

**Pull request merged**